### PR TITLE
VestingEscrow for CC

### PR DESCRIPTION
**Pull request by hinkypunks**

### CODE REVIEW

File: contracts/RetroDistribution/VestingEscrowProxy.sol [hunk 1]
```diff
@@ -16,105 +16,129 @@ contract VestingEscrowProxy is Initializable, ProxyReentrancyGuard, ProxyOwned,
...
     }
 
-    function setStartTime(uint256 _startTime) external onlyOwner {
-        startTime = _startTime;
+    function disableClaim(address _recipient) external onlyOwner {
+        disabledAt[_recipient] = block.timestamp;
+        emit DisableClaim(_recipient);
+    }
+
+    function enableClaim(address _recipient) external onlyOwner {
```
Comment (grujawork) [inline contracts/RetroDistribution/VestingEscrowProxy.sol:117]: check zero address

File: contracts/RetroDistribution/VestingEscrowProxy.sol [hunk 2]
```diff
@@ -16,105 +16,129 @@ contract VestingEscrowProxy is Initializable, ProxyReentrancyGuard, ProxyOwned,
...
+        
         IERC20Upgradeable(token).safeTransfer(msg.sender, claimable);
+
+        totalClaimed[msg.sender] = totalClaimed[msg.sender] + claimable;
         emit Claim(msg.sender, claimable);
     }
 
-    function setStartTime(uint256 _startTime) external onlyOwner {
-        startTime = _startTime;
+    function disableClaim(address _recipient) external onlyOwner {
```
Comment (grujawork) [inline contracts/RetroDistribution/VestingEscrowProxy.sol:112]: check zero address

File: contracts/RetroDistribution/VestingEscrowProxy.sol [hunk 3]
```diff
@@ -16,105 +16,129 @@ contract VestingEscrowProxy is Initializable, ProxyReentrancyGuard, ProxyOwned,
...
+    }
+
+    function setStartTime(address _recipient, uint _startTime) external onlyOwner {
+        startTime[_recipient] = _startTime;
+        emit StartTimeChanged(_recipient, _startTime);
     }
 
-    function setEndTime(uint256 _endTime) external onlyOwner {
-        endTime = _endTime;
+    function setEndTime(address _recipient, uint _endTime) external onlyOwner {
```
Comment (grujawork) [inline contracts/RetroDistribution/VestingEscrowProxy.sol:127]: check _endTime (must be in the future maybe, etc.)

File: contracts/RetroDistribution/VestingEscrowProxy.sol [hunk 4]
```diff
@@ -16,105 +16,129 @@ contract VestingEscrowProxy is Initializable, ProxyReentrancyGuard, ProxyOwned,
     using SafeERC20Upgradeable for IERC20Upgradeable;
 
     address public token;
-    uint256 public startTime;
```
Comment (danijelthales) [inline contracts/RetroDistribution/VestingEscrowProxy.sol:19]: This should be a new contract 'VestingEscrowCCs.sol'

File: contracts/RetroDistribution/VestingEscrowProxy.sol [hunk 5]
```diff
@@ -16,105 +16,129 @@ contract VestingEscrowProxy is Initializable, ProxyReentrancyGuard, ProxyOwned,
...
 
-        initialLockedSupply = initialLockedSupply.add(_totalAmount);
+        initialLockedSupply = initialLockedSupply + _totalAmount;
     }
 
-    function _totalVestedOf(address _recipient, uint256 _time) internal view returns (uint256) {
-        uint256 start = startTime;
-        uint256 end = endTime;
-        uint256 locked = initialLocked[_recipient];
+    function _totalVestedOf(address _recipient, uint _time) internal view returns (uint) {
```
Comment (danijelthales) [inline contracts/RetroDistribution/VestingEscrowProxy.sol:63]: why was the change from uint256 to uint needed?

File: contracts/RetroDistribution/VestingEscrowProxy.sol [hunk 6]
```diff
@@ -16,105 +16,129 @@ contract VestingEscrowProxy is Initializable, ProxyReentrancyGuard, ProxyOwned,
...
 
-    function _totalVested() internal view returns (uint256) {
-        uint256 start = startTime;
-        uint256 end = endTime;
-        uint256 locked = initialLockedSupply;
-
-        if (block.timestamp < start) {
-            return 0;
+    function _totalVested() internal view returns (uint totalVested) {
+        for(uint i = 0; i < recipients.length; i++) {
```
Comment (danijelthales) [inline contracts/RetroDistribution/VestingEscrowProxy.sol:73]: and this definitely works even when we got 50 CCs?

**Review by danijelthales — APPROVED** [review]

**Review by danijelthales — DISMISSED** [review]

**Review by grujawork — APPROVED** [review]

**Pull request merged**